From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Damien Caliste <dcaliste@free.fr>
Date: Tue, 21 Nov 2023 09:51:51 +0100
Subject: [PATCH] Adjust for Qt 5.6

Revert "Fix bundled zlib detection"
This reverts commit 7e2579c1fbfc077f3d95eceed5c79236752acb87.

qtConfig() not available before 5.8.0 so reverting the change until
qt gets updated.

Revert "Use QRandomGenerator instead of qrand"
This reverts commit 9c8962c1942eec471d495b3d11436087e7d46cdd.
---
 src/libraries/qmfclient/libaccounts_p.cpp     |  4 ++--
 src/libraries/qmfclient/mailkeyimpl_p.h       |  6 ++---
 src/libraries/qmfclient/mailsortkeyimpl_p.h   |  4 ++--
 src/libraries/qmfclient/qmailkeyargument.h    |  2 +-
 src/libraries/qmfclient/qmailmessage.cpp      | 22 +++++++++++++++--
 src/libraries/qmfclient/qmailmessageset.h     | 24 +++++++++++++++++++
 .../qmfstoragemanager/qmfstoragemanager.cpp   |  3 +--
 src/plugins/messageservices/imap/imap.pro     | 14 ++++++++---
 .../messageservices/smtp/smtpclient.cpp       | 12 ++++++++--
 tests/tst_smtp/tst_smtp.cpp                   | 18 +++++++-------
 10 files changed, 83 insertions(+), 26 deletions(-)

diff --git a/src/libraries/qmfclient/libaccounts_p.cpp b/src/libraries/qmfclient/libaccounts_p.cpp
index 79645a03..390e73e8 100644
--- a/src/libraries/qmfclient/libaccounts_p.cpp
+++ b/src/libraries/qmfclient/libaccounts_p.cpp
@@ -569,7 +569,7 @@ QMailAccount LibAccountManager::account(const QMailAccountId &id) const
     if ((static_cast<uint>(account->valueAsUInt64(QLatin1String("lastSynchronized")))) == 0) {
         result.setLastSynchronized(QMailTimeStamp());
     } else {
-        result.setLastSynchronized(QMailTimeStamp(QDateTime::fromSecsSinceEpoch(static_cast<uint>(account->valueAsUInt64(QLatin1String("lastSynchronized"))))));
+        result.setLastSynchronized(QMailTimeStamp(QDateTime::fromTime_t(static_cast<uint>(account->valueAsUInt64(QLatin1String("lastSynchronized"))))));
     }
 
     result.setIconPath(account->valueAsString(QLatin1String("iconPath")));
@@ -924,7 +924,7 @@ bool LibAccountManager::updateSharedAccount(QMailAccount *account,
         sharedAccount->setValue(QLatin1String("fullName"), account->fromAddress().name());
         sharedAccount->setValue(QLatin1String("emailAliases"), toStringList(account->fromAliases()));
         if (account->lastSynchronized().isValid()) {
-            sharedAccount->setValue(QLatin1String("lastSynchronized"), static_cast<quint64>(account->lastSynchronized().toLocalTime().toSecsSinceEpoch()));
+            sharedAccount->setValue(QLatin1String("lastSynchronized"), static_cast<quint64>(account->lastSynchronized().toLocalTime().toTime_t()));
         } else {
             sharedAccount->setValue(QLatin1String("lastSynchronized"), quint64(0));
         }
diff --git a/src/libraries/qmfclient/mailkeyimpl_p.h b/src/libraries/qmfclient/mailkeyimpl_p.h
index 5170f2a9..6f9f53b2 100644
--- a/src/libraries/qmfclient/mailkeyimpl_p.h
+++ b/src/libraries/qmfclient/mailkeyimpl_p.h
@@ -292,16 +292,16 @@ void MailKeyImpl<Key>::deserialize(Stream &stream)
     combiner = static_cast<QMailKey::Combiner>(i);
     stream >> negated;
 
-    qsizetype s;
+    int s;
     stream >> s;
-    for (qsizetype j = 0; j < s; ++j) {
+    for (int j = 0; j < s; ++j) {
         Argument a;
         a.deserialize(stream);
         arguments.append(a);
     }
 
     stream >> s;
-    for (qsizetype j = 0; j < s; ++j) {
+    for (int j = 0; j < s; ++j) {
         Key subKey;
         subKey.deserialize(stream);
         subKeys.append(subKey);
diff --git a/src/libraries/qmfclient/mailsortkeyimpl_p.h b/src/libraries/qmfclient/mailsortkeyimpl_p.h
index d6171f62..95cd5e9f 100644
--- a/src/libraries/qmfclient/mailsortkeyimpl_p.h
+++ b/src/libraries/qmfclient/mailsortkeyimpl_p.h
@@ -133,9 +133,9 @@ template<typename Key>
 template <typename Stream>
 void MailSortKeyImpl<Key>::deserialize(Stream &stream)
 {
-    qsizetype i = 0;
+    int i = 0;
     stream >> i;
-    for (qsizetype j = 0; j < i; ++j) {
+    for (int j = 0; j < i; ++j) {
         Argument a;
         a.deserialize(stream);
         _arguments.append(a);
diff --git a/src/libraries/qmfclient/qmailkeyargument.h b/src/libraries/qmfclient/qmailkeyargument.h
index 0d776be7..d1f0ec8e 100644
--- a/src/libraries/qmfclient/qmailkeyargument.h
+++ b/src/libraries/qmfclient/qmailkeyargument.h
@@ -162,7 +162,7 @@ public:
         {
             clear();
 
-            qsizetype v = 0;
+            int v = 0;
             stream >> v;
             for (int i = 0; i < v; ++i) {
                 QVariant var;
diff --git a/src/libraries/qmfclient/qmailmessage.cpp b/src/libraries/qmfclient/qmailmessage.cpp
index 37aee1c2..d1fc9c56 100644
--- a/src/libraries/qmfclient/qmailmessage.cpp
+++ b/src/libraries/qmfclient/qmailmessage.cpp
@@ -50,7 +50,6 @@
 #include <QFile>
 #include <QFileInfo>
 #include <QRegularExpression>
-#include <QRandomGenerator>
 #include <QDataStream>
 #include <QTextStream>
 #include <QTextCodec>
@@ -8003,6 +8002,25 @@ bool QMailMessagePrivate::hasRecipients() const
     return false;
 }
 
+static uint currentTimeValue()
+{
+    return QDateTime::currentDateTime().toTime_t();
+}
+
+static bool seedRng()
+{
+    qsrand(currentTimeValue());
+    return true;
+}
+
+static int randomNumber()
+{
+    static bool initialised = seedRng();
+    Q_UNUSED(initialised)
+
+    return qrand();
+}
+
 static QByteArray gBoundaryString;
 
 void QMF_EXPORT setQMailMessageBoundaryString(const QByteArray &boundary)
@@ -8019,7 +8037,7 @@ static QByteArray boundaryString(const QByteArray &hash)
         return gBoundaryString;
 
     // Formulate a boundary that is very unlikely to clash with the content
-    return boundaryLeader + "qmf:" + QByteArray::number(QRandomGenerator::global()->generate()) + hash.toBase64() + boundaryTrailer;
+    return boundaryLeader + "qmf:" + QByteArray::number(randomNumber()) + hash.toBase64() + boundaryTrailer;
 }
 
 template <typename F>
diff --git a/src/libraries/qmfclient/qmailmessageset.h b/src/libraries/qmfclient/qmailmessageset.h
index a4545ce5..4ea9c5ac 100644
--- a/src/libraries/qmfclient/qmailmessageset.h
+++ b/src/libraries/qmfclient/qmailmessageset.h
@@ -79,7 +79,11 @@ protected:
 
 private:
     // _D version and 'priv' to avoid collision with QObject::d_ptr
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailMessageSetContainer)
+#else
+    Q_DECLARE_PRIVATE_D(QMailMessageSetContainer::priv.data(), QMailMessageSetContainer)
+#endif
     Q_DISABLE_COPY(QMailMessageSetContainer)
 
     friend class QMailMessageSet;
@@ -117,7 +121,11 @@ protected:
     virtual void init();
 
 private:
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailMessageSet)
+#else
+    Q_DECLARE_PRIVATE_D(QMailMessageSet::priv.data(), QMailMessageSet)
+#endif
 };
 
 
@@ -155,7 +163,11 @@ protected:
     QMailFolderKey folderKey() const;
 
 private:
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailFolderMessageSet)
+#else
+    Q_DECLARE_PRIVATE_D(QMailFolderMessageSet::priv.data(), QMailFolderMessageSet)
+#endif
 };
 
 
@@ -194,7 +206,11 @@ protected:
     QMailFolderKey rootFolderKey() const;
 
 private:
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailAccountMessageSet)
+#else
+    Q_DECLARE_PRIVATE_D(QMailAccountMessageSet::priv.data(), QMailAccountMessageSet)
+#endif
 };
 
 
@@ -228,7 +244,11 @@ protected:
     void resyncState() override;
 
 private:
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailFilterMessageSet)
+#else
+    Q_DECLARE_PRIVATE_D(QMailFilterMessageSet::priv.data(), QMailFilterMessageSet)
+#endif
 };
 
 
@@ -317,7 +337,11 @@ protected:
     virtual void updated(QMailMessageSet *child);
 
 private:
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 9, 0))
     Q_DECLARE_PRIVATE_D(priv, QMailMessageSetModel)
+#else
+    Q_DECLARE_PRIVATE_D(QMailMessageSetModel::priv.data(), QMailMessageSetModel)
+#endif
 
     friend class QMailMessageSetContainer;
     friend class QMailMessageSet;
diff --git a/src/plugins/contentmanagers/qmfstoragemanager/qmfstoragemanager.cpp b/src/plugins/contentmanagers/qmfstoragemanager/qmfstoragemanager.cpp
index c3167088..06c333fd 100644
--- a/src/plugins/contentmanagers/qmfstoragemanager/qmfstoragemanager.cpp
+++ b/src/plugins/contentmanagers/qmfstoragemanager/qmfstoragemanager.cpp
@@ -43,7 +43,6 @@
 #include <QFile>
 #include <QtPlugin>
 #include <QUrl>
-#include <QRandomGenerator>
 
 #if defined(Q_OS_WIN)
 #include <windows.h>
@@ -79,7 +78,7 @@ QString randomString(int length)
 
     int i = 0;
     while (length--) {
-        int r = QRandomGenerator::global()->generate() % 62;
+        int r = qrand() % 62;
         r += 48;
         if (r>57) r+=7;
         if (r>90) r+=6;
diff --git a/src/plugins/messageservices/imap/imap.pro b/src/plugins/messageservices/imap/imap.pro
index 54addffe..e9c8dc35 100644
--- a/src/plugins/messageservices/imap/imap.pro
+++ b/src/plugins/messageservices/imap/imap.pro
@@ -38,10 +38,18 @@ SOURCES += imapclient.cpp \
            serviceactionqueue.cpp \
            idlenetworksession.cpp
 
-qtConfig(system-zlib) {
-    QMAKE_USE_PRIVATE += zlib
+!contains(QT_CONFIG, system-zlib) {
+    INCLUDEPATH += $$[QT_INSTALL_HEADERS]/QtZlib
+    DEFINES += QT_QMF_HAVE_ZLIB
+} else:packagesExist(zlib) {
+    CONFIG += link_pkgconfig
+    PKGCONFIG += zlib
+    DEFINES += QT_QMF_HAVE_ZLIB
+} else:macx:exists("/usr/include/zlib.h") {
+    LIBS += -lz
+    DEFINES += QT_QMF_HAVE_ZLIB
 } else {
-    QT_PRIVATE += zlib-private
+    warning("IMAP COMPRESS capability requires zlib")
 }
 
 # hack longstream work without the private/ on include
diff --git a/src/plugins/messageservices/smtp/smtpclient.cpp b/src/plugins/messageservices/smtp/smtpclient.cpp
index 9be40d99..e7ba6121 100644
--- a/src/plugins/messageservices/smtp/smtpclient.cpp
+++ b/src/plugins/messageservices/smtp/smtpclient.cpp
@@ -42,7 +42,6 @@
 #include <QDir>
 #include <QHostInfo>
 #include <QNetworkInterface>
-#include <QRandomGenerator>
 #include <QRegularExpression>
 #include <QSslSocket>
 
@@ -58,9 +57,18 @@ Q_LOGGING_CATEGORY(lcSMTP, "org.qt.messageserver.smtp", QtWarningMsg)
 // Only this many bytes is queued to be sent at a time.
 #define SENDING_BUFFER_SIZE 5000
 
+static bool initialiseRng()
+{
+    qsrand(QDateTime::currentDateTime().toUTC().toTime_t());
+    return true;
+}
+
 static QByteArray messageId(const QByteArray& domainName, quint32 addressComponent)
 {
-    quint32 randomComponent(QRandomGenerator::global()->generate());
+    static bool rngInitialised(initialiseRng());
+    Q_UNUSED(rngInitialised)
+
+    quint32 randomComponent(static_cast<quint32>(qrand()));
     quint32 timeComponent(QDateTime::currentDateTimeUtc().toMSecsSinceEpoch() / 1000);
 
     return ('<' +
diff --git a/tests/tst_smtp/tst_smtp.cpp b/tests/tst_smtp/tst_smtp.cpp
index 5784e7c0..e7f197d5 100644
--- a/tests/tst_smtp/tst_smtp.cpp
+++ b/tests/tst_smtp/tst_smtp.cpp
@@ -94,9 +94,9 @@ void tst_SmtpClient::test_connection()
     QVERIFY(completed.wait());
 
     QCOMPARE(updateStatus.count(), 3);
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("DNS lookup"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("DNS lookup"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
 }
 
 void tst_SmtpClient::test_auth()
@@ -115,9 +115,9 @@ void tst_SmtpClient::test_auth()
     QVERIFY(!completed.wait(250)); // Fails with wrong credentials
 
     QCOMPARE(updateStatus.count(), 3);
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("DNS lookup"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("DNS lookup"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
 
     smtp.setSmtpAuthentication(QMail::PlainMechanism);
     QVERIFY(QMailStore::instance()->updateAccountConfiguration(&config));
@@ -126,9 +126,9 @@ void tst_SmtpClient::test_auth()
     QVERIFY(!completed.wait(250)); // Fails with wrong credentials
 
     QCOMPARE(updateStatus.count(), 3);
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("DNS lookup"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
-    QCOMPARE(updateStatus.takeFirst().first(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("DNS lookup"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
+    QCOMPARE(updateStatus.takeFirst().first().toString(), QString::fromLatin1("Connected"));
 }
 
 QTEST_MAIN(tst_SmtpClient)
